src/renderer/renderElement.ts

```diff
@@ -93,6 +94,49 @@ export interface ExcalidrawElementWithCanvas {
   boundTextElementVersion: number | null;
 }
 
+const cappedElementCanvasSize = (
+  element: NonDeletedExcalidrawElement,
+  zoom: Zoom,
+): {
+  width: number;
+  height: number;
+  scale: NormalizedZoomValue;
+} => {
+  const sizelimit = 16777216; //iOS Safari canvas size limit
+  const widthHeightLimit = 32767; //based on developer.mozilla.org
+  const padding = getCanvasPadding(element);
+  let scale = zoom.value;
+
+  const [x1, y1, x2, y2] = getElementAbsoluteCoords(element);
+  const elementWidth =
+    isLinearElement(element) || isFreeDrawElement(element)
+      ? distance(x1, x2)
+      : element.width;
+  const elementHeight =
+    isLinearElement(element) || isFreeDrawElement(element)
+      ? distance(y1, y2)
+      : element.height;
+
+  let width = elementWidth * window.devicePixelRatio + padding * 2;
+  let height = elementHeight * window.devicePixelRatio + padding * 2;
+
+  //recalculate scale to ensure width and height are within
+  //the allowed maximums
+  if (width > widthHeightLimit || height > widthHeightLimit) {
```

Comment:
```suggestion
  if (width * scale > widthHeightLimit || height * scale > widthHeightLimit) {
```

should check the post-resize dimensions